text/1026-net-module.rst

```diff
@@ -0,0 +1,200 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2024-08-14
+    Authors: Scott Trinh <scott@edgedb.com>
+
+=============================
+RFC: EdgeDB Networking Module
+=============================
+
+This RFC outlines the addition of a networking module to EdgeDB’s
+standard library, providing functionality for making HTTP requests and
+sending SMTP messages.
+
+Motivation
+==========
+
+The EdgeDB Auth extension (ext::auth) currently sends emails directly to
+configured SMTP services at certain times in various authentication
+ceremonies and workflows:
+
+-  Sign up with email (sends email verification email)
+-  Password reset (sends password reset email)
+-  Magic link sign up (sends magic link email)
+-  Magic link sign in (sends magic link email)
+
+To allow full customization of emails and provide hooks into
+authentication events generally, we need to support sending webhook
+requests to the application directly. This would enable applications to
+trigger their own logic for sending fully customized and
+internationalized emails, SMS, log analytics, update usage billing
+records, etc.
+
+Specification
+=============
+
+Module Name
+-----------
+
+The new module will be named ``net``.
+
+Functions
+---------
+
+The module will provide the following functions:
+
+1. HTTP Request Function:
+
+.. code-block:: edgeql
+
+   scalar type net::HttpMethod extending std::enums<`GET`, POST, PUT, `DELETE`, PATCH, HEAD, OPTIONS>;
+
+   function net::http_request(
+       url: str,
+       named only body: optional str,
+       named only method: net::HttpMethod = net::HttpMethod::GET,
+       named only headers: optional array<tuple<name: str, value: str>>
+   ) -> net::HttpResponse;
+
+2. SMTP Send Function:
+
+.. code-block:: edgeql
+
+   function net::smtp_send(
+       smtp_url: str,
+       named only from: multi str,
+       named only to: multi str,
+       named only subject: str,
+       named only text: optional str,
+       named only html: optional str,
+   ) -> net::SmtpResponse
+
+ResponseState enum
+------------------
+
+Since network requests are asynchronous, the response state will be
+represented by an enum:
+
+.. code-block:: edgeql
+
+   scalar type net::ResponseState extending std::enums<Pending, InProgress, Complete, Failed>;
```

Comment:
There could be some difference between the status of a response and the state of the requesting task, and such a difference may affect how we retry the request.

A task could have such states:

* `Pending`
* `InProgress`
* `Complete`
* `Failed`
* `Error`

I added an `Error` state for non-retryable failures, like a bad request, too many retries, timed out in receiving response (server may have accepted the request without us knowing). In this case, we can only retry on `Failed` state (transitioning to `InProgress`), or certain `Complete` cases depending on the response status below.

Meanwhile, a response may have different statuses (based on the HTTP status):

* `Success` (200-ish)
* `ClientError` (400-ish)
* `ServerError` (500-ish)

By default, I think we shouldn't retry on client errors, but it may be safe to always retry on server errors. Still, this would be very much domain-specific (the remote may not follow the HTTP semantics as we assumed), so ultimately, we may have to let users decide when to retry. However, perhaps we don't need an additional `HttpStatus` now (there is already the `status` field), especially for the `try_request` MVP. I just feel wrong mixing this response status with the task state. Like the task state can be shared between `http`, `smtp` or friends, but the response status is protocol-specific.

Reply:
> There could be some difference between the status of a response and the state of the requesting task, and such a difference may affect how we retry the request.

Yeah, I meant to communicate that the `ResponseState` was indeed the state of the _task_, but I didn't have the concept of `Task` initially.

> I added an Error state for non-retryable failures, like a bad request, too many retries, timed out in receiving response (server may have accepted the request without us knowing). In this case, we can only retry on Failed state (transitioning to InProgress), or certain Complete cases depending on the response status below.

Yeah, this is an interesting idea. It feels a little like the `Task` would be driving a lot of the logic in this case, which I'm not sure I think is a good idea. I'd rather have the thing that knows about retrying own that logic and know how to inspect a `Task<HttpResponse, TaskError>` and retry the `Task`. I think this is the fundamental disagreement that I was having with @1st1 also. To me, these basic methods for making requests should _not_ have any logic and just know protocol-specific things, rather than "business logic" like things (including retrying, concurrency, etc).

> However, perhaps we don't need an additional HttpStatus now (there is already the status field), especially for the try_request MVP.

Yeah, I agree with this. The `task.response.status` should just be the status code returned by the server.

> I just feel wrong mixing this response status with the task state. Like the task state can be shared between http, smtp or friends, but the response status is protocol-specific.

Absolutely, and this "state" part was meant to be the `Task.state`, so yeah, I'll refactor.

Reply:
@fantix ⚡ [`3c588ec` (#95)](https://github.com/edgedb/rfcs/pull/95/commits/3c588ecf17359b0ba2de07aa004d65d0f6d11093)